matchtree.go

```diff
@@ -321,10 +282,61 @@ func (t *fileNameMatchTree) prepare(doc uint32) {
 	t.child.prepare(doc)
 }
 
-func (t *substrMatchTree) prepare(nextDoc uint32) {
-	t.matchIterator.prepare(nextDoc)
+func (t *substrMatchTree) prepare(doc uint32) {
+	t.matchIterator.prepare(doc)
 	t.current = t.matchIterator.candidates()
 	t.contEvaluated = false
+
+	var fileStart uint32
+	if doc > 0 {
+		fileStart = t.fileEndRunes[doc-1]
+	}
+
+	var sections []DocumentSection
+	if len(t.sections) > 0 {
+		most := t.fileEndSymbol[len(t.fileEndSymbol)-1]
+		if most == uint32(len(t.sections)) {
+			sections = t.sections[t.fileEndSymbol[doc]:t.fileEndSymbol[doc+1]]
+		} else {
+			for t.secID < uint32(len(t.sections)) && t.sections[t.secID].Start < fileStart {
+				t.secID++
+			}
+
+			fileEnd, symbolEnd := t.fileEndRunes[doc], t.secID
+			for symbolEnd < uint32(len(t.sections)) && t.sections[symbolEnd].Start < fileEnd {
+				symbolEnd++
+			}
+
+			sections = t.sections[t.secID:symbolEnd]
+		}
+	}
+
+	secIdx := 0
+	trimmed := t.current[:0]
```

Comment:
This still needs some cleaning up. I moved this code from symbolSubstrMatchTree. However, since we don't filter but rather enrich, we could do without "trimmed" and just mutate t.current.